src/Featurizers/Components/DocumentStatisticsEstimator.h

```diff
@@ -246,31 +279,276 @@ inline DocumentStatisticsAnnotationData::DocumentStatisticsAnnotationData(Freque
+                                                                                    std::float_t totalNumDocumentsFloat,
+                                                                                    nonstd::optional<std::uint32_t> topKTerms) {
 
-            return decorator;
+    using FrequencyMap = Details::DocumentStatisticsTrainingOnlyPolicy::FrequencyMap;
+
+    //trim by minDf and maxDf
+    if (minDf > 0.0f || maxDf < 1.0f) {
+        FrequencyMap::iterator termFrequencyIter = termFrequency.begin();
+        while (termFrequencyIter != termFrequency.end()) {
+            std::float_t const freq = termFrequencyIter->second / totalNumDocumentsFloat;
+
+            if (freq < minDf || freq > maxDf) {
+                termFrequencyIter = termFrequency.erase(termFrequencyIter);
+            } else {
+                ++termFrequencyIter;
+            }
+        }
+    }
+
+    //trim by topKTerms
+    if (!topKTerms.has_value() || termFrequency.empty() || *topKTerms >= termFrequency.size())
+        return termFrequency;
+
+    std::uint32_t const & topKTermsValue(*topKTerms);
+
+    std::tuple<std::uint32_t, long long> targetPack(
+        [&termFrequency, &topKTermsValue]() -> std::tuple<std::uint32_t, long long> {
+            std::vector<std::uint32_t> frequencyVector;
+            frequencyVector.reserve(termFrequency.size());
+
+            std::transform(
+                termFrequency.begin(),
+                termFrequency.end(),
+                std::back_inserter(frequencyVector),
+                [](std::pair<std::string, std::uint32_t> const & kv) {
+                    return kv.second;
+                }
+            );
+
+            //partial sort, ave O(n)
+            std::nth_element(
+                frequencyVector.begin(),
+                frequencyVector.begin() + static_cast<int>(topKTermsValue) - 1,
+                frequencyVector.end(),
+                std::greater<std::uint32_t>()
+            );
+
+            std::uint32_t targetValue = frequencyVector[topKTermsValue - 1];
+            long long remainingTargetValue = std::count(
+                frequencyVector.begin(),
+                frequencyVector.begin() + static_cast<int>(topKTermsValue),
+                targetValue
+            );
+
+            return std::make_tuple<std::uint32_t, long long>(
+                std::move(targetValue),
+                std::move(remainingTargetValue)
+            );
         }()
+    );
+
+    std::uint32_t targetValue = std::get<0>(targetPack);
+    long long remainingTargetValue = std::get<1>(targetPack);
+
+    FrequencyMap::iterator termFrequencyIter = termFrequency.begin();
+    while (termFrequencyIter != termFrequency.end()) {
+        if (termFrequencyIter->second < targetValue) {
+            termFrequencyIter = termFrequency.erase(termFrequencyIter);
+            continue;
+        } else if (termFrequencyIter->second == targetValue) {
```

Comment:
In the first if clause, you are continuing, however, in this one you are not. I'd suggest writing this loop so that you leverage continue if the iterator has been updated due to an erase and only increment the iterator in one place.

Reply:
there's place for no increment, I'm not sure how to use loop here. I'll remain the original one for now